File: quicksort/quicksort.js
function qsort(arr) {
  if (!arr.length) return [];

  var pivot = arr.splice(Math.random() * arr.length | 0, 1)[0];
  var left = arr.filter(function (el) { return el <= pivot; });
  var right = arr.filter(function (el) { return el > pivot; });

  return qsort(left).concat(pivot).concat(qsort(right));
}

// input handling

var data = '';
process.stdin.resume();
process.stdin.on('data', function (chunk) { data += chunk; });

process.stdin.on('end', function () {
  var arr = data
    .split('\n')
    .slice(0, -1)
    .map(function (el) { return parseInt(el, 10); });

  qsort(arr).forEach(function (el) { console.log(el); });
});

File: quicksort/quicksort.go
package main

import (
	"fmt"
	"io/ioutil"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

func filter(s []int, fn func(int) bool) []int {
	p := []int{}
	for _, v := range s {
		if fn(v) {
			p = append(p, v)
		}
	}
	return p
}

func qsort(arr []int) []int {
	if len(arr) == 0 {
		return []int{}
	}

	pix := rand.Intn(len(arr))
	p := arr[pix]
	arr = append(arr[:pix], arr[pix+1:]...)
	left := filter(arr, func(i int) bool {
		return i < p
	})
	right := filter(arr, func(i int) bool {
		return i >= p
	})

	return append(qsort(left), append([]int{p}, qsort(right)...)...)
}

func main() {
	rand.Seed(time.Now().UnixNano())

	bytes, err := ioutil.ReadAll(os.Stdin)

	if err != nil {
		panic(err)
	}

	arr := []int{}
	for _, v := range strings.Split(string(bytes), "\n") {
		n, err := strconv.Atoi(v)

		if err == nil {
			arr = append(arr, n)
		}
	}

	for _, n := range qsort(arr) {
		fmt.Println(n)
	}
}
